Hit Fused Addmm Path for Nd tensors

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106740
* __->__ #106433
* #106739

Speeds up convnext bfloat16 inference by 35%., and timm bfloat16 inference average by `.5%`


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @anijain2305

eellison committed:
[test]

[ghstack-poisoned]

eellison committed:
Update on "[test]"




[ghstack-poisoned]

eellison committed:
Update on "[test]"




[ghstack-poisoned]

eellison committed:
Update on "Hit Fused Addmm Path for 4d tensors"


Speeds up convnext bfloat16 inference by 35%. Doing full benchmark run.


[ghstack-poisoned]

vadimkantorov commented:
Related: https://github.com/pytorch/pytorch/issues/39661 https://github.com/pytorch/pytorch/pull/74070

Seems this has been done a few times in different contexts :)

vadimkantorov commented on aten/src/ATen/native/Linear.cpp:
```diff
@@ -75,19 +79,20 @@ Tensor linear(const Tensor& input, const Tensor& weight, const c10::optional<Ten
     return at::addmm(*bias, input, weight.t());
   }
-  if (input.dim() == 3 && bias->defined() && !input.is_xla()) {
+  if ((input_dim == 3 || input_dim == 4) && bias->defined() && !input.is_xla()) {
```
what is the motivation behind only fusing for 3-dim and 4-dim inputs? any contig inputs should work fine, right? (as can reshape without any perf hit)

eellison replied:
yea, no motivation, just didnt have real world models to test on. I can update to hit it for all dimensions.

vadimkantorov replied:
in my personal view, the less hidden differences/surprises like this, the better...

eellison committed:
Update on "Hit Fused Addmm Path for 4d tensors"


Speeds up convnext bfloat16 inference by 35%., and timm bfloat16 inference average by `.5%`


[ghstack-poisoned]

eellison committed:
Update on "Hit Fused Addmm Path for 4d tensors"


Speeds up convnext bfloat16 inference by 35%., and timm bfloat16 inference average by `.5%`


[ghstack-poisoned]

eellison committed:
Update on "Hit Fused Addmm Path for Nd tensors"


Speeds up convnext bfloat16 inference by 35%., and timm bfloat16 inference average by `.5%`


[ghstack-poisoned]

eellison committed:
Update on "Hit Fused Addmm Path for Nd tensors"


Speeds up convnext bfloat16 inference by 35%., and timm bfloat16 inference average by `.5%`


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

Commit:
Update on "Hit Fused Addmm Path for Nd tensors"


Speeds up convnext bfloat16 inference by 35%., and timm bfloat16 inference average by `.5%`


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Hit Fused Addmm Path for Nd tensors"


Speeds up convnext bfloat16 inference by 35%., and timm bfloat16 inference average by `.5%`


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

eellison reviewed (commented):
cc @alexmsettle this is breaking `test_inductor_sequence_nr`. would you take a look ?

alexmsettle commented on test/dynamo/test_aot_autograd.py:
```diff
@@ -810,39 +810,40 @@ def _prepare_model_args():
+1|aten.native_batch_norm_backward.default|
+0|aten.convolution_backward.default|
+"""
             ),
```
@eellison - this change didn't fix the failing test?  It looks like your change set changed the aten ops which would change the fx listing of aten ops.  I would have recommended just changing the expected output like you did here.  Let me know if you need me to dig into this further.
